src/lib.rs

```diff
@@ -136,42 +124,61 @@ impl<'a> Fragment<'a> {
                     .and_then(|j| {
                         fragment.get(i..i+j)
                     })
-            })
+            });
+        Ok(temp)
     }
 
     // Get virtual cubes from one fragment
-    fn get_virtual_cubes(fragment: &'a str) -> Option<&'a str> {
-        fragment.find(VIRTUALCUBE_TAG_OPEN)
+    fn get_virtual_cubes(fragment: &'a str) -> Result<Option<&'a str>> {
+        let temp = fragment.find(VIRTUALCUBE_TAG_OPEN)
             .and_then(|i| {
                 fragment[i..]
                     .find(SCHEMA_TAG_CLOSE)
                     .or(Some(fragment.len()-i)) // eof
                     .and_then(|j| {
                         fragment.get(i..i+j)
                     })
-            })
+            });
+        Ok(temp)
     }
 
-    pub fn process_fragment(fragment: &'a str) -> Fragment<'a> {
+    pub fn process_fragment(fragment: &'a str) -> Result<Fragment<'a>> {
         // TODO make this work with string parse fn?
 
-        let schema_name = Fragment::get_schema_name(fragment);
-        let shared_dims = Fragment::get_shared_dims(fragment);
-        let cubes = Fragment::get_cubes(fragment);
-        let virtual_cubes = Fragment::get_virtual_cubes(fragment);
-
-        Fragment {
+        let schema_name;
+        let shared_dims;
+        let cubes;
+        let virtual_cubes;
+
+        match Fragment::get_schema_name(fragment) {
```

Comment:
This pattern is good! Something to simplify it is the `?` operator. So you should be able to use `get_schema_name(fragment)?`.

If you look here, you've replicated the code almost exactly :) https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html